# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
import numpy as np
import pandas as pd
from sklearn.svm import SVR
import statsmodels.api as sm
import matplotlib.pyplot as plt
from sklearn.feature_selection import RFE
from sklearn.linear_model import LinearRegression 
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, mean_squared_error

# Challenge 1 - Explore the Scikit-Learn Datasets

Before starting to work on our own datasets, let's first explore the datasets that are included in this Python library. These datasets have been cleaned and formatted for use in ML algorithms.

First, we will load the diabetes dataset. Do this in the cell below by importing the datasets and then loading the dataset  to the `diabetes` variable using the `load_diabetes()` function.

In [2]:
# Your code here:
from sklearn.datasets import load_diabetes
data = load_diabetes()

Let's explore this variable by looking at the different attributes. Do this by looking at the `keys()` of this variable.

In [3]:
# Your code here:
data.keys()

dict_keys(['data', 'target', 'frame', 'DESCR', 'feature_names', 'data_filename', 'target_filename'])

The next step is to read the description of the dataset. Print the description in the cell below using the `DESCR` attribute of the `diabetes` variable

In [4]:
# Your code here:
data.DESCR

'.. _diabetes_dataset:\n\nDiabetes dataset\n----------------\n\nTen baseline variables, age, sex, body mass index, average blood\npressure, and six blood serum measurements were obtained for each of n =\n442 diabetes patients, as well as the response of interest, a\nquantitative measure of disease progression one year after baseline.\n\n**Data Set Characteristics:**\n\n  :Number of Instances: 442\n\n  :Number of Attributes: First 10 columns are numeric predictive values\n\n  :Target: Column 11 is a quantitative measure of disease progression one year after baseline\n\n  :Attribute Information:\n      - age     age in years\n      - sex\n      - bmi     body mass index\n      - bp      average blood pressure\n      - s1      tc, T-Cells (a type of white blood cells)\n      - s2      ldl, low-density lipoproteins\n      - s3      hdl, high-density lipoproteins\n      - s4      tch, thyroid stimulating hormone\n      - s5      ltg, lamotrigine\n      - s6      glu, blood sugar level\n\nNo

What are the variables in this dataset according to the description? List them in the markdown cell below

This dataset is composed of 442 rows and 10 columns
- Each **row** in the dataset corresponds to a patient
- Each **column** in the dataset corresponds to ten different measured variables related to each patient (age, sex and bmi) or to physiological variables that have been measured in each of them.
    - `age`: in years
    - `sex`: F/M 
    - `bmi`: body mass index
    - `bp` : average blood pressure
    - `s1` : T-Cells (a type of white blood cells)
    - `s2` : low-density lipoproteins (bad cholesterol)
    - `s3` : high-density lipoproteins (good cholesterol)
    - `s4` : thyroid stimulating hormone
    - `s5` : lamotrigine
    - `s6` : blood sugar level  
 

#### Enter your answer here:




Now explore the data. Scikit-learn typically takes in 2D numpy arrays as input (though pandas dataframes are also accepted). In the cell below find the shape of the numpy array contained in the data portion of the diabetes variable.

In [5]:
# Your code here:
data.target.shape

(442,)

# Challenge 2 - Perform Supervised Learning on the Dataset

#### The data has already been split to predictor and response variables. The response variable is in the `target` portion of the variable. 

Given this information, let's apply what we have previously learned about linear regression and apply the algorithm to the diabetes dataset. In the cell below, import the linear regression class from sklearn. 

In [6]:
x = data.data
y = data.target

Initialize the model in the variable `diabetes_model`

In [7]:
# Your code here:
diabetes_model = LinearRegression(fit_intercept = True) #fit_intercept to calculate the intercept of the model if is True

In the cell below, fit the model and print the intercept and coefficients of the model. 

In [8]:
# Your code here:
diabetes_model.fit(x,y)

LinearRegression()

In [9]:
diabetes_model.coef_

array([ -10.01219782, -239.81908937,  519.83978679,  324.39042769,
       -792.18416163,  476.74583782,  101.04457032,  177.06417623,
        751.27932109,   67.62538639])

Each coefficients represent the relationship between the given feature and the target. As a general rule the coefficients can be interpreted as: 

> For every one-unit increase in [X variable], the [y variable] increases by [coefficient] when all other variables are held constant.


In this way, these coefficients are telling us if our predictor variables are positively or negatively related to our response variable and what the magnitude of that relationship is.

[scikit-learn](https://scikit-learn.org/stable/auto_examples/inspection/plot_linear_model_coefficient_interpretation.html)  
[coefficients](https://towardsdatascience.com/interpreting-coefficients-in-linear-and-logistic-regression-6ddf1295f6f1)

In [10]:
 diabetes_model.intercept_

152.1334841628965

The intercept is the expected mean value of Y when all the predictor variables are 0. 

# Bonus Challenge 1 - Conduct a Hypothesis Test on the Model

Once we have generated a linear model, we can test each coefficient using a t-test to see whether the confidence interval for the variable contains zero. We can also perform an overall F test to check whether at least one coefficient is significantly different from zero. 

Refer to the resource in this [link](https://onlinecourses.science.psu.edu/stat501/node/297/) for more details and perform the t-tests for the model above. Additionally, interpret the results and list coefficients are significantly different from zero.


Hint: use the statsmodels package.

Your result should look similar to this:

![ols](../ols-results.png)

In [11]:
# Your code here:
model = sm.OLS(y, x).fit()

In [12]:
predictions = model.predict(x)
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                      y   R-squared (uncentered):                   0.106
Model:                            OLS   Adj. R-squared (uncentered):              0.085
Method:                 Least Squares   F-statistic:                              5.100
Date:                Tue, 23 Feb 2021   Prob (F-statistic):                    4.72e-07
Time:                        21:30:19   Log-Likelihood:                         -2873.9
No. Observations:                 442   AIC:                                      5768.
Df Residuals:                     432   BIC:                                      5809.
Df Model:                          10                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
x1           -10.0122    179.967     -0.056      0.956    -363.731     343.707
x2          -239.8191    184.404     -1.301      0.194    -602.259     122.621
x3           519.8398    200.401      2.594      0.010     125.957     913.723
x4           324.3904    197.053      1.646      0.100     -62.912     711.692
x5          -792.1842   1255.066     -0.631      0.528   -3258.979    1674.611
x6           476.7458   1021.183      0.467      0.641   -1530.360    2483.851
x7           101.0446    640.155      0.158      0.875   -1157.162    1359.251
x8           177.0642    486.370      0.364      0.716    -778.882    1133.010
x9           751.2793    517.774      1.451      0.148    -266.391    1768.950
x10           67.6254    198.747      0.340      0.734    -323.005     458.256
==============================================================================
Omnibus:                        1.506   Durbin-Watson:                   0.223
Prob(Omnibus):                  0.471   Jarque-Bera (JB):                1.404
Skew:                           0.017   Prob(JB):                        0.496
Kurtosis:                       2.726   Cond. No.                         21.7
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

[OLS](https://www.statsmodels.org/dev/examples/notebooks/generated/ols.html)

# Challenge 2 - Peform Supervised Learning on a Pandas Dataframe

Now that we have looked at data that has been formatted for scikit-learn, let's look at data that we will need to format ourselves.

In the next cell, load the `auto-mpg.csv` file included in this folder and assign it to a variable called `auto`.

In [13]:
# Your code here:
auto = pd.read_csv('../auto-mpg.csv')

Look at the first 5 rows using the `head()` function:

In [14]:
# Your code here:
auto.head()

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   
3  16.0          8         304.0        150.0    3433          12.0   
4  17.0          8         302.0        140.0    3449          10.5   

   model_year                       car_name  
0          70  \t"chevrolet chevelle malibu"  
1          70          \t"buick skylark 320"  
2          70         \t"plymouth satellite"  
3          70              \t"amc rebel sst"  
4          70                \t"ford torino"

Evaluate the data to ensure that all numeric columns are correctly detected as such by pandas. If a column is misclassified as object, coerce it to numeric.

**¡¡¡DF.INFO() TOP TOP!!!** gives us all the information about null types of each column of the dataframe

All the columns are well classified

In [18]:
# Your code here
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 398 entries, 0 to 397
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           398 non-null    float64
 1   cylinders     398 non-null    int64  
 2   displacement  398 non-null    float64
 3   horse_power   392 non-null    float64
 4   weight        398 non-null    int64  
 5   acceleration  398 non-null    float64
 6   model_year    398 non-null    int64  
 7   car_name      398 non-null    object 
dtypes: float64(4), int64(3), object(1)
memory usage: 25.0+ KB


What is the newest model year and the oldest model year?

In [20]:
# Your code here: 
oldest  = auto["model_year"].max()
newest = auto["model_year"].min()
print(f'The oldest model is {oldest}, the newest model is {newest}')

The oldest model is 82, the newest model is 70


Check the dataset for missing values and remove all rows containing at least one missing value.

In [21]:
# Your code here:
auto.isna().sum(axis = 0) # There aer 6 NaN in the column horse_power
auto[pd.isnull(auto["horse_power"])] # get the rows with NaN

mpg  cylinders  displacement  horse_power  weight  acceleration  \
32   25.0          4          98.0          NaN    2046          19.0   
126  21.0          6         200.0          NaN    2875          17.0   
330  40.9          4          85.0          NaN    1835          17.3   
336  23.6          4         140.0          NaN    2905          14.3   
354  34.5          4         100.0          NaN    2320          15.8   
374  23.0          4         151.0          NaN    3035          20.5   

     model_year                  car_name  
32           71            \t"ford pinto"  
126          74         \t"ford maverick"  
330          80  \t"renault lecar deluxe"  
336          80    \t"ford mustang cobra"  
354          81           \t"renault 18i"  
374          82        \t"amc concord dl"

In [22]:
auto = auto.dropna() # remove the rows with NaN

Find the frequency table for the `cylinders` column using the `value_counts()` function. How many possible values of cylinders are there?

In [23]:
# Your code here:
values = auto["cylinders"].value_counts().count()
print(f'There are {values} possible values of cylinders')

There are 5 possible values of cylinders


We would like to generate a linear regression model that will predict mpg. To do this, first drop the `car_name` column since it does not contain any quantitative data. Next separate the dataframe to predictor and response variables. Separate those into test and training data with 80% of the data in the training set and the remainder in the test set. 

Assign the predictor and response training data to `X_train` and `y_train` respectively. Similarly, assign the predictor and response test data to `X_test` and `y_test`.

In [24]:
#DROP COLUMNS CAR_NAME
auto.drop(["car_name"], axis = 1).head(3)

mpg  cylinders  displacement  horse_power  weight  acceleration  \
0  18.0          8         307.0        130.0    3504          12.0   
1  15.0          8         350.0        165.0    3693          11.5   
2  18.0          8         318.0        150.0    3436          11.0   

   model_year  
0          70  
1          70  
2          70

In [25]:
#First of all assign the the response and predictor variables:
X = auto.loc[:,"cylinders":"model_year"]
Y = auto["mpg"]

In [26]:
# Your code here:
X_train, X_test, Y_train, Y_test  = train_test_split(X, Y, test_size= 0.2, random_state = None)

Now we will the dataset that we processed and peform linear regression on this data to predict the mpg for each vehicle. Initialize the model in the cell below.

In [27]:
# Your code here:
car_model = LinearRegression()

Next, fit the model in the cell below.

In [28]:
# Your code here:
car_model.fit(X_train, Y_train)
accuracy = car_model.score(X_train, Y_train) # Check accuracy / variance

# Challenge 3 - Evaluate the Model

the r squared score of a model tells us how much variation is explained by the model. In a typical dataset, most observations differ from the mean. When we create a model, we are trying to generate an equation that will tell us by how much each observation will differ from the mean. Obviously, the vast majority of models are not perfect. They can only predict some of the variation from the mean but not all of it. We attribute the rest of the difference between the actual value and the mean to random error. We would like random error to explain the as little as possible of the variation. This is why the r squared score is an important metric.

In the next cell, compute the r squared score of the model. Do this by first computing the predicted values and assign them to `y_pred`.

In [29]:
# Your code here:
y_pred = car_model.predict(X_train)

In [30]:
print('Test R squared:', accuracy)

Test R squared: 0.798300540215551


In [31]:
train_pred = pd.DataFrame({'Observed': Y_train, 'Predicted': y_pred}) 
train_pred

Observed  Predicted
353      33.0  32.241684
80       22.0  24.252735
84       27.0  25.879694
48       18.0  18.902252
97       18.0  20.191023
..        ...        ...
313      28.0  28.582709
76       18.0  20.773971
207      20.0  22.598391
102      26.0  27.954244
289      16.9  16.551974

[313 rows x 2 columns]

#### Our next step is to evaluate the model using the test data. We would like to ensure that our model is not overfitting the data. This means that our model will not be able to generalize well outside of the training data.

In the cell below, use the model to generate the predicted values for the training data and assign them to `y_test_pred`. Compute the r squared score for the test data by comparing the oberserved `y_test` data and the predicted `y_test_pred`.

In [32]:
# Your code here:
y_test_pred  = car_model.predict(X_test)
accuracy_test = car_model.score(X_test, Y_test)
print('Test R squared:', accuracy_test)

Test R squared: 0.8387786360803532


In [33]:
train_pred = pd.DataFrame({'Observed': Y_test, 'Predicted': y_test_pred}) 
train_pred

Observed  Predicted
206      26.5  26.463464
73       13.0  13.027236
271      23.2  26.546764
319      31.3  29.431799
113      21.0  23.766751
..        ...        ...
160      17.0  16.946688
187      17.5  15.192597
37       18.0  17.805095
199      20.0  19.323769
258      20.6  22.399063

[79 rows x 2 columns]

# Challenge 4 - Improve the Model Fit

While the most common way to improve the fit of a model is by using regularization, there are other simpler ways to improve model fit. The first is to create a simpler model. The second is to increase the train sample size.

Let us start with the easier option and increase our train sample size to 90% of the data. Create a new test train split and name the new predictors and response variables `X_train09`, `X_test09`, `y_train09`, `y_test09`.

In [34]:
# Your code here:
X_train09, X_test09, y_train09, y_test09 = train_test_split(X, Y, test_size = 0.2, random_state = None)

Initialize a new model. Name this model `auto_model09`. Fit the model to the new sample data.

In [35]:
# Your code here:
auto_model09 = LinearRegression()
auto_model09.fit(X_train09, y_train09)

LinearRegression()

In [36]:
y_pred09 = auto_model09.predict(X_train09)

In [37]:
accuracy_train09 = auto_model09.score(X_train09, y_train09)
print('Test R squared:', accuracy_train09)

Test R squared: 0.7946188374572827


Compute the predicted values and r squared score for our new model and new sample data.

In [38]:
# Your code here:
train_pred09 = pd.DataFrame({'Observed': y_train09, 'Predicted': y_pred09}) 
train_pred09

Observed  Predicted
238      33.5  30.122186
111      18.0  26.412114
375      36.0  34.389448
48       18.0  18.693157
153      18.0  19.649442
..        ...        ...
386      25.0  27.866676
368      27.0  30.307279
183      25.0  28.719715
269      30.9  29.438652
59       23.0  26.445212

[313 rows x 2 columns]

Compute the r squared score for the smaller test set. Is there an improvement in the test r squared?

In [39]:
# Your code here:
y_test_pred09  = auto_model09.predict(X_test09)

In [40]:
accuracy_test09 = auto_model09.score(X_test09, y_test09)
print('Test R squared:', accuracy_test09)

Test R squared: 0.8573574499836044


In [41]:
test_pred09 = pd.DataFrame({'Observed': y_test09, 'Predicted': y_test_pred09}) 
test_pred09

Observed  Predicted
77       22.0  23.866504
302      34.5  30.813439
391      36.0  31.452241
90       12.0   8.281874
247      39.4  31.180840
..        ...        ...
28        9.0   7.551116
122      24.0  22.917666
9        15.0  13.182789
91       13.0  11.324316
189      15.5  16.238919

[79 rows x 2 columns]

# Bonus Challenge 2 - Backward Elimination 

The main way to produce a simpler linear regression model is to reduce the number of variables used in the model. In scikit-learn, we can do this by using recursive feature elimination. You can read more about RFE [here](https://scikit-learn.org/stable/modules/generated/sklearn.feature_selection.RFE.html).

In the next cell, we will import RFE

Follow the documentation and initialize an RFE model using the `auto_model` linear regression model. Set `n_features_to_select=3`

From the documentation cited above:  

"Given an external estimator that assigns weights to features (e.g., the coefficients of a linear model), the goal of recursive feature elimination (RFE) is to select features by recursively considering smaller and smaller sets of features. First, the estimator is trained on the initial set of features and the importance of each feature is obtained either through any specific attribute or callable. Then, the least important features are pruned from current set of features. That procedure is recursively repeated on the pruned set until the desired number of features to select is eventually reached."


**More information:**  
[RFE](https://machinelearningmastery.com/feature-selection-in-python-with-scikit-learn/)

In [42]:
# create a base classifier used to evaluate a subset of attributes
estimator = SVR(kernel="linear")

In [43]:
# create the RFE model and select 3 attributes
rfe = RFE(estimator, n_features_to_select=3)

Fit the model and print the ranking

In [44]:
# Your code here:
rfe.fit(X, Y)
rfe.ranking_


array([1, 3, 2, 4, 1, 1])

Each value in the array indicate the attribute ranking for each of the variables included on the model. Therefore seleted features are assigned rank 1

>cylinders = 1  
>displacement = 3  
>horse_power = 2  
>weight = 4  
>acceleration = 1  
>model_year = 1  

[ResultsInterpretationRFE](https://medium.com/@aneesha/recursive-feature-elimination-with-scikit-learn-3a2cbdf23fb7)

Feature importance is ranked from most important (1) to least important (4). Generate a model with the three most important features. The features correspond to variable names. For example, feature 1 is `cylinders` and feature 2 is `displacement`.

Perform a test-train split on this reduced column data and call the split data `X_train_reduced`, `X_test_reduced`, `y_test_reduced`, `y_train_reduced`. Use an 80% split.

In [53]:
# Your code here:
x1 = auto[["cylinders", "displacement"]]
y1 = auto['mpg']
X_train_reduced, X_test_reduced, y_train_reduced, y_test_reduced = train_test_split(x1, y1, test_size = 0.2, random_state = None)

Generate a new model called `auto_model_reduced` and fit this model. Then proceed to compute the r squared score for the model. Did this cause an improvement in the r squared score?

In [54]:
# Your code here: 
auto_model_reduced = LinearRegression()
auto_model_reduced.fit(X_train_reduced, y_train_reduced)

LinearRegression()

In [55]:
y_pred_reduced = auto_model_reduced.predict(X_train_reduced)

In [66]:
accuracy_train_reduced = auto_model_reduced.score(X_train_reduced, y_train_reduced)
print('Test R squared:', accuracy_train09 , 'Test R squared_Reduced model:', accuracy_test09)

Test R squared: 0.7946188374572827 Test R squared_Reduced model: 0.8573574499836044


When we compare the two R2 of our models, we see that the second model that we have evaluated after the RFE has a higher R2, therefore our model is explaining more variation of our data.